### ISSUE TITLE
How to get the same performance as paper on LFW, CFP-FP, AgeDB-30?

### ISSUE THREAD

**Issue by yzhHoward** [IrvingMeng/MagFace#17]
Hello! According to the paper, I trained on MS1MV2, extract the feature directly by iResNet50 and use the code in [ArcFace](https://github.com/deepinsight/insightface/tree/master/recognition/arcface_torch) when validating on LFW, CFP-FF, CFP-FP and AgeDB-30.
However, I evaluate the model on every epoch and only get best accuracy of 0.98283 on LFW, 0.97814 on CFP-FF, 0.77314 on CFP-FP, and 0.89300 on AgeDB-30.
Here is the training log on MS1MV2 [output.log](https://github.com/IrvingMeng/MagFace/files/6698278/output.log). I changed the learning rate and batch size to 1.25 times the original while training.

Is there anything wrong?

**Comment by IrvingMeng** [comment]
Hi, the log seems all right.  The issue may be caused by incorrect features.  Could you provide more details about the configs which are missed in the log? 

Several common mistakes:
1.  inconsistent structures for training and testing
2. inconsistent pre-processing step. This repo uses (x-0)/1 while the code in insightface uses (x-127.5)/127.5.
3. fail to load the checkpoint

**Comment by yzhHoward** [comment]
Thank you for replying! The following is my code (From [validation](https://github.com/yzhHoward/MagFace/tree/main))
Code for extracting feature:
```python
class SoftmaxBuilder(nn.Module):
    def __init__(self, args):
        super(SoftmaxBuilder, self).__init__()
        self.features = load_features(args)
        self.fc = MagLinear(args.embedding_size,
                            args.last_fc_size,
                            scale=args.arc_scale)

        self.l_margin = args.l_margin
        self.u_margin = args.u_margin
        self.l_a = args.l_a
        self.u_a = args.u_a

    def _margin(self, x):
        """generate adaptive margin
        """
        margin = (self.u_margin-self.l_margin) / \
            (self.u_a-self.l_a)*(x-self.l_a) + self.l_margin
        return margin

    def forward(self, x, target=None):
        x = self.features(x)
        # logits, x_norm = self.fc(x, self._margin, self.l_a, self.u_a)
        return x
```

Code for loading model and validating:
```python
def do_val(model, criterion, optimizer, args):
    image_size = [112, 112]
    val_dataset = 'lfw,cfp_ff,cfp_fp,agedb_30,calfw,cplfw'
    print('start validation')
    # print('validating ' + ver_name_list[i])
    for name in val_dataset.split(','):
        path = os.path.join('/data/jupyter/faces_emore/', name + ".bin")
        if os.path.exists(path):
            print('loading.. ', name)
            data_set = load_bin(path, image_size)
        results = []
        for i in range(1, 26):
            print(i)
            if i < 10:
                checkpoint = torch.load('0000' + str(i) + '.pth')
            else:
                checkpoint = torch.load('000' + str(i) + '.pth')
            # checkpoint = torch.load('00025.pth', map_location=torch.device('cpu'))
            from collections import OrderedDict
            new_state_dict = OrderedDict()
            for k, v in checkpoint['state_dict'].items():
                param_name = k[7:] # remove `module.`
                new_state_dict[param_name] = v
            model.load_state_dict(new_state_dict)
            optimizer.load_state_dict(checkpoint['optimizer'])
            model.eval()
            with torch.no_grad():
                acc1, std1, acc2, std2, xnorm, embeddings_list = test(
                    data_set, model, args.batch_size, 10)
                print('[%s]XNorm: %f' % (name, xnorm))
                print('[%s]Accuracy-Flip: %1.5f+-%1.5f' % (name, acc2, std2))
                results.append(acc2)
        print('Max of [%s] is %1.5f' % (name, np.max(results)))
```
Code from ArcFace:
```python
import datetime
import os
import pickle

import mxnet as mx
import numpy as np
import sklearn
import torch
from mxnet import ndarray as nd
from scipy import interpolate
from sklearn.decomposition import PCA
from sklearn.model_selection import KFold


class LFold:
    def __init__(self, n_splits=2, shuffle=False):
        self.n_splits = n_splits
        if self.n_splits > 1:
            self.k_fold = KFold(n_splits=n_splits, shuffle=shuffle)

    def split(self, indices):
        if self.n_splits > 1:
            return self.k_fold.split(indices)
        else:
            return [(indices, indices)]


def calculate_roc(thresholds,
                  embeddings1,
                  embeddings2,
                  actual_issame,
                  nrof_folds=10,
                  pca=0):
    assert (embeddings1.shape[0] == embeddings2.shape[0])
    assert (embeddings1.shape[1] == embeddings2.shape[1])
    nrof_pairs = min(len(actual_issame), embeddings1.shape[0])
    nrof_thresholds = len(thresholds)
    k_fold = LFold(n_splits=nrof_folds, shuffle=False)

    tprs = np.zeros((nrof_folds, nrof_thresholds))
    fprs = np.zeros((nrof_folds, nrof_thresholds))
    accuracy = np.zeros((nrof_folds))
    indices = np.arange(nrof_pairs)

    if pca == 0:
        diff = np.subtract(embeddings1, embeddings2)
        dist = np.sum(np.square(diff), 1)

    for fold_idx, (train_set, test_set) in enumerate(k_fold.split(indices)):
        if pca > 0:
            print('doing pca on', fold_idx)
            embed1_train = embeddings1[train_set]
            embed2_train = embeddings2[train_set]
            _embed_train = np.concatenate((embed1_train, embed2_train), axis=0)
            pca_model = PCA(n_components=pca)
            pca_model.fit(_embed_train)
            embed1 = pca_model.transform(embeddings1)
            embed2 = pca_model.transform(embeddings2)
            embed1 = sklearn.preprocessing.normalize(embed1)
            embed2 = sklearn.preprocessing.normalize(embed2)
            diff = np.subtract(embed1, embed2)
            dist = np.sum(np.square(diff), 1)

        # Find the best threshold for the fold
        acc_train = np.zeros((nrof_thresholds))
        for threshold_idx, threshold in enumerate(thresholds):
            _, _, acc_train[threshold_idx] = calculate_accuracy(
                threshold, dist[train_set], actual_issame[train_set])
        best_threshold_index = np.argmax(acc_train)
        for threshold_idx, threshold in enumerate(thresholds):
            tprs[fold_idx, threshold_idx], fprs[fold_idx, threshold_idx], _ = calculate_accuracy(
                threshold, dist[test_set],
                actual_issame[test_set])
        _, _, accuracy[fold_idx] = calculate_accuracy(
            thresholds[best_threshold_index], dist[test_set],
            actual_issame[test_set])

    tpr = np.mean(tprs, 0)
    fpr = np.mean(fprs, 0)
    return tpr, fpr, accuracy


def calculate_accuracy(threshold, dist, actual_issame):
    predict_issame = np.less(dist, threshold)
    tp = np.sum(np.logical_and(predict_issame, actual_issame))
    fp = np.sum(np.logical_and(predict_issame, np.logical_not(actual_issame)))
    tn = np.sum(
        np.logical_and(np.logical_not(predict_issame),
                       np.logical_not(actual_issame)))
    fn = np.sum(np.logical_and(np.logical_not(predict_issame), actual_issame))

    tpr = 0 if (tp + fn == 0) else float(tp) / float(tp + fn)
    fpr = 0 if (fp + tn == 0) else float(fp) / float(fp + tn)
    acc = float(tp + tn) / dist.size
    return tpr, fpr, acc


def calculate_val(thresholds,
                  embeddings1,
                  embeddings2,
                  actual_issame,
                  far_target,
                  nrof_folds=10):
    assert (embeddings1.shape[0] == embeddings2.shape[0])
    assert (embeddings1.shape[1] == embeddings2.shape[1])
    nrof_pairs = min(len(actual_issame), embeddings1.shape[0])
    nrof_thresholds = len(thresholds)
    k_fold = LFold(n_splits=nrof_folds, shuffle=False)

    val = np.zeros(nrof_folds)
    far = np.zeros(nrof_folds)

    diff = np.subtract(embeddings1, embeddings2)
    dist = np.sum(np.square(diff), 1)
    indices = np.arange(nrof_pairs)

    for fold_idx, (train_set, test_set) in enumerate(k_fold.split(indices)):

        # Find the threshold that gives FAR = far_target
        far_train = np.zeros(nrof_thresholds)
        for threshold_idx, threshold in enumerate(thresholds):
            _, far_train[threshold_idx] = calculate_val_far(
                threshold, dist[train_set], actual_issame[train_set])
        if np.max(far_train) >= far_target:
            f = interpolate.interp1d(far_train, thresholds, kind='slinear')
            threshold = f(far_target)
        else:
            threshold = 0.0

        val[fold_idx], far[fold_idx] = calculate_val_far(
            threshold, dist[test_set], actual_issame[test_set])

    val_mean = np.mean(val)
    far_mean = np.mean(far)
    val_std = np.std(val)
    return val_mean, val_std, far_mean


def calculate_val_far(threshold, dist, actual_issame):
    predict_issame = np.less(dist, threshold)
    true_accept = np.sum(np.logical_and(predict_issame, actual_issame))
    false_accept = np.sum(
        np.logical_and(predict_issame, np.logical_not(actual_issame)))
    n_same = np.sum(actual_issame)
    n_diff = np.sum(np.logical_not(actual_issame))
    # print(true_accept, false_accept)
    # print(n_same, n_diff)
    val = float(true_accept) / float(n_same)
    far = float(false_accept) / float(n_diff)
    return val, far


def evaluate(embeddings, actual_issame, nrof_folds=10, pca=0):
    # Calculate evaluation metrics
    thresholds = np.arange(0, 4, 0.01)
    embeddings1 = embeddings[0::2]
    embeddings2 = embeddings[1::2]
    tpr, fpr, accuracy = calculate_roc(thresholds,
                                       embeddings1,
                                       embeddings2,
                                       np.asarray(actual_issame),
                                       nrof_folds=nrof_folds,
                                       pca=pca)
    thresholds = np.arange(0, 4, 0.001)
    val, val_std, far = calculate_val(thresholds,
                                      embeddings1,
                                      embeddings2,
                                      np.asarray(actual_issame),
                                      1e-3,
                                      nrof_folds=nrof_folds)
    return tpr, fpr, accuracy, val, val_std, far

@torch.no_grad()
def load_bin(path, image_size):
    if os.path.exists(path + '.pkl'):
        return pickle.load(open(path + '.pkl', 'rb'))
    try:
        with open(path, 'rb') as f:
            bins, issame_list = pickle.load(f)  # py2
    except UnicodeDecodeError as e:
        with open(path, 'rb') as f:
            bins, issame_list = pickle.load(f, encoding='bytes')  # py3
    data_list = []
    for flip in [0, 1]:
        data = torch.empty((len(issame_list) * 2, 3, image_size[0], image_size[1]))
        data_list.append(data)
    for idx in range(len(issame_list) * 2):
        _bin = bins[idx]
        img = mx.image.imdecode(_bin)
        if img.shape[1] != image_size[0]:
            img = mx.image.resize_short(img, image_size[0])
        img = nd.transpose(img, axes=(2, 0, 1))
        for flip in [0, 1]:
            if flip == 1:
                img = mx.ndarray.flip(data=img, axis=2)
            data_list[flip][idx][:] = torch.from_numpy(img.asnumpy())
        if idx % 1000 == 0:
            print('loading bin', idx)
    print(data_list[0].shape)
    pickle.dump([data_list, issame_list], open(path + '.pkl', 'wb'))
    print('dump data to ' + path + '.pkl')
    return data_list, issame_list

@torch.no_grad()
def test(data_set, backbone, batch_size, nfolds=10):
    print('testing verification..')
    data_list = data_set[0]
    issame_list = data_set[1]
    embeddings_list = []
    time_consumed = 0.0
    for i in range(len(data_list)):
        data = data_list[i]
        embeddings = None
        ba = 0
        while ba < data.shape[0]:
            bb = min(ba + batch_size, data.shape[0])
            count = bb - ba
            _data = data[bb - batch_size: bb]
            time0 = datetime.datetime.now()
            img = ((_data / 255) - 0.5) / 0.5
            # img = _data
            img = img.cuda(non_blocking=True)
            # compute output
            _embeddings = backbone(img)
            _embeddings = _embeddings.detach().cpu().numpy()
            time_now = datetime.datetime.now()
            diff = time_now - time0
            time_consumed += diff.total_seconds()
            if embeddings is None:
                embeddings = np.zeros((data.shape[0], _embeddings.shape[1]))
            embeddings[ba:bb, :] = _embeddings[(batch_size - count):, :]
            ba = bb
        embeddings_list.append(embeddings)

    _xnorm = 0.0
    _xnorm_cnt = 0
    for embed in embeddings_list:
        for i in range(embed.shape[0]):
            _em = embed[i]
            _norm = np.linalg.norm(_em)
            _xnorm += _norm
            _xnorm_cnt += 1
    _xnorm /= _xnorm_cnt

    embeddings = embeddings_list[0].copy()
    embeddings = sklearn.preprocessing.normalize(embeddings)
    acc1 = 0.0
    std1 = 0.0
    embeddings = embeddings_list[0] + embeddings_list[1]
    embeddings = sklearn.preprocessing.normalize(embeddings)
    print(embeddings.shape)
    print('infer time', time_consumed)
    _, _, accuracy, val, val_std, far = evaluate(embeddings, issame_list, nrof_folds=nfolds)
    acc2, std2 = np.mean(accuracy), np.std(accuracy)
    return acc1, std1, acc2, std2, _xnorm, embeddings_list
```
Besides, it seems strange that the results become worse (acc 0.7) when applying (x-0)/1 during pre-procesing.

**Comment by yzhHoward** [comment]
Well, I validate the offical released model (magface_iresnet50_MS1MV2_dp.pth)...
When applying (x-0)/1 during pre-procesing, the result is around acc 0.7.
When applying the code in ArcFace, the result on LFW is 0.91217, CFP-FF 0.92314, CFP-FP 0.71586, AgeDB-30 0.84717, CALFW 0.85800, CPLFW 0.70600...
Maybe the validating method is not the same as ArcFace referred in the paper?

**Comment by IrvingMeng** [comment]
> Well, I validate the offical released model (magface_iresnet50_MS1MV2_dp.pth)...
> When applying (x-0)/1 during pre-procesing, the result is around acc 0.7.
> When applying the code in ArcFace, the result on LFW is 0.91217, CFP-FF 0.92314, CFP-FP 0.71586, AgeDB-30 0.84717, CALFW 0.85800, CPLFW 0.70600...
> Maybe the validating method is not the same as ArcFace referred in the paper?

The validating method should be the same as my evaluation codes are also modified from InsightFace :)

I still believe the low performances are caused by incorrect features. I have uploaded my evaluation codes and datasets for lfw/cfp/agedb (details in readme). You may be able to find out the reason by comparing the differences. If things goes well, you can get:

![image](https://user-images.githubusercontent.com/10399303/123042776-e5dac700-d429-11eb-9e99-e52dce1c8ac7.png)

**Comment by yzhHoward** [comment]
It really works! Thanks a lot!
